stdlib/immutablearray.gr

```diff
@@ -0,0 +1,935 @@
+
+  let numToAppend = Number.min(len2, branchingFactor - len1)
+  let toAppend = if (numToAppend < len2) {
+    Array.slice(0, numToAppend, array2)
+  } else {
+    array2
+  }
+  let numNotAppended = len1 + len2 - branchingFactor
+
+  (Array.append(array1, toAppend), numNotAppended)
+}
+
+let log32floor = num => {
+  let mut val = -1
+  let mut num = num
+  while (num > 0) {
+    val += 1
+    num = num >> branchingBits
+  }
+  val
+}
+
+/**
+ * Attempts to retrieve the element from the array at the specified index.
+ * A negative index is treated as an offset from the end of the array.
+ *
+ * @param index: The index to access
+ * @param array: The array to access
+ * @returns `Some(element)` containing the element from the array at the given index or `None` if the index is out of the array bounds
+ * 
+ * @example getOpt(1, fromList([1, 2, 3])) == Some(2)
+ * @example getOpt(-1, fromList([1, 2, 3])) == Some(3)
+ * @example getOpt(4, fromList([1, 2, 3])) == None
+ *
+ * @since v0.5.5
+ */
+export let getOpt = (index, array) => {
+  let index = wrapNegativeIndex(array.length, index)
+
+  let rec getInner = (shift, node) => {
+    let pos = index >> shift & bitmask
+    match (node[pos]) {
+      Tree(subTree) => getInner(shift - branchingBits, subTree),
+      Leaf(vals) => vals[index & bitmask],
+    }
+  }
+
+  let { length, shift, root, tail } = array
+  if (index < 0 || index >= length) {
+    None
+  } else if (index >= tailIndex(length)) {
+    Some(tail[index & bitmask])
+  } else {
+    Some(getInner(shift, root))
+  }
+}
+
+/**
+ * Retrieves the element from the array at the specified index.
+ * A negative index is treated as an offset from the end of the array.
+ *
+ * @param index: The index to access
+ * @param array: The array to access
+ * @returns The element from the array
+ * 
+ * @example get(1, fromList([1, 2, 3, 4, 5])) == 2
+ *
+ * @since v0.5.5
+ */
+export let get = (index, array) => Option.unwrap(getOpt(index, array))
+
+/**
+ * Creates a new array in which the element at the specified index is set to a
+ * new value. A negative index is treated as an offset from the end of the array.
+ *
+ * @param index: The index to update
+ * @param value: The value to store
+ * @param array: The array to update
+ * @returns A new array containing the new element at the given index
+ * 
+ * @example set(1, 9, fromList([1, 2, 3, 4, 5])) == fromList([1, 9, 3, 4, 5])
+ *
+ * @since v0.5.5
+ */
+export let set = (index, value, array) => {
+  let index = wrapNegativeIndex(array.length, index)
+
+  let rec setInner = (shift, node) => {
+    let pos = index >> shift & bitmask
+    let newVal = match (node[pos]) {
+      Tree(subTree) => {
+        Tree(setInner(shift - branchingBits, subTree))
+      },
+      Leaf(vals) => {
+        Leaf(arraySetCopy(index & bitmask, value, vals))
+      },
+    }
+    arraySetCopy(pos, newVal, node)
+  }
+
+  let { length, shift, root, tail } = array
+  if (index < 0 || index >= length) {
+    array
```

Comment:
Should we fail in this case?

Reply:
I think so, otherwise it's failing silently.

Reply:
Sure, should it return a result or throw an exception?

Reply:
I just went with throwing `IndexOutOfBounds` for now since it matches regular array behavior. On this topic, do you think that `getOpt` should exist at all, and if so should it be added to `Array` as well?

Reply:
Personally, I like having getOpt available.

Reply:
Sounds good. On second thought, should I actually rename it to `getOption`? The standard library generally doesn't have a lot of contractions in names.

Reply:
Since `get` returns a value instead of the ImmutableArray, why did you choose to make it fallible and then have `getOpt`?

Reply:
In order to mirror the behavior for `Array`, where `get` would just return the value at that index and throw `IndexOutOfBounds` otherwise. I figured I'd also add `getOpt` to be used where safety is desired. Also why would `get` return the array?

Reply:
Honestly, `Array.get` doesn't return an option because it is implemented as `arr[idx]` which throws on index out of bounds. It'd be great if writing `arr[idx]` was the "throws" version and `Array.get` were the safe version.

> Also why would get return the array?

Sorry, I meant in opposition to `set` which should throw with IndexOutOfBounds because it needs to return the ImmutableArray (and returning `Result<ImmutableArray, Err>` is a terrible API).

Reply:
Ok so are you saying that `getOpt` should be removed and `get` should always return `Option` (changing this behavior for `Array` too)? I still think it would be nice to have a "quick and dirty" way to extract a value from an array though rather than always having to jump through the hoop of unwrapping or handle the `None` case.

I suppose an alternative solution would be to somehow allow the bracket syntax to be used for immutable arrays (either somehow make this an allowable way to construct operators or a built-in syntax for immutable arrays)?

Reply:
I really hate putting a "return type" in your function name, e.g. `getOption`. It's just bad design.

Reply:
What about `safeGet`?

Reply:
I think we remove the version that returns an option until we decide to handle for mutable arrays.